comp/forwarder/defaultforwarder/blocked_endpoints.go

```diff
@@ -84,22 +115,67 @@ func (e *blockedEndpoints) recover(endpoint string) {
 	if knownBlock, ok := e.errorPerEndpoint[endpoint]; ok {
 		b = knownBlock
 	} else {
-		b = &block{}
+		b = &block{
+			state: Closed,
+		}
 	}
 
-	b.nbError = e.backoffPolicy.DecError(b.nbError)
-	b.until = time.Now().Add(e.getBackoffDuration(b.nbError))
+	switch b.state {
+	case Closed:
+		// Nothing to do, we are already closed and running smoothly.
+	case HalfOpen:
+		// The test worked, we can ease off
+		b.nbError = e.backoffPolicy.DecError(b.nbError)
+		if b.nbError == 0 {
+			b.state = Closed
+		} else {
+			b.until = TimeNow().Add(e.getBackoffDuration(b.nbError))
+			b.state = Open
+		}
+	case Open:
+		// If we are open and a successful transaction came through, we
+		// can't be sure if it was sent before the errored transaction or
+		// after, so we will ignore this.
+	}
 
 	e.errorPerEndpoint[endpoint] = b
 }
 
 func (e *blockedEndpoints) isBlock(endpoint string) bool {
 	e.m.RLock()
-	defer e.m.RUnlock()
 
-	if b, ok := e.errorPerEndpoint[endpoint]; ok && time.Now().Before(b.until) {
-		return true
+	if b, ok := e.errorPerEndpoint[endpoint]; ok {
+		switch b.state {
+		case HalfOpenTest:
+			// We have already sent the single transactions to test if the endpoint is now up
+			e.m.RUnlock()
+			return true
+		case HalfOpen:
+			e.m.RUnlock()
+			e.m.Lock()
+			// Send a single transaction to test the endpoint.
+			b.state = HalfOpenTest
+			e.m.Unlock()
+			return false
+		case Open:
+			if TimeNow().Before(b.until) {
+				e.m.RUnlock()
+				return true
+			} else {
+				// Upgrade to a full lock so we can move
+				// to HalfOpen and send this transaction
+				// to test the endpoint.
+				e.m.RUnlock()
+				e.m.Lock()
```

Comment:
Same as above, multiple callers may reach this point while state is `Closed` and the timer is expired, and will proceed to make the request.